### PR TITLE
Default layout strategy for catalogs

### PR DESCRIPTION
**Pull request by thomas-maschler**
**Related Issue(s):**

- https://github.com/stac-utils/pystac-client/issues/241

**Description:**

This PR enables setting a default layout strategy for a catalog. Users can either set the default layout strategy
when instantiating the catalog or defining a fall back strategy when subclassing Catalog.

Together with #1294, this PR is part of setting the foundation for enabling API transactions using `pystac-client`.

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `scripts/test`)
- [x] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

File: pystac/catalog.py [hunk 1]
```diff
@@ -167,6 +171,9 @@ class Catalog(STACObject):
     a canonical format.
     """
 
+    _layout_strategy: HrefLayoutStrategy = BestPracticesLayoutStrategy()
```
Comment (jsignell) [inline pystac/catalog.py:174]: So the idea is that pystac will be able to override this in the `Client` class?

File: pystac/catalog.py [hunk 2]
```diff
@@ -128,6 +128,10 @@ class Catalog(STACObject):
             catalog's self link's HREF.
         catalog_type : Optional catalog type for this catalog. Must
             be one of the values in :class:`~pystac.CatalogType`.
+        strategy : The layout strategy to use for setting the
```
Comment (jsignell) [inline pystac/catalog.py:131]: See comment above about choosing `strategy` vs `layout_strategy` for the kwarg.

**Review by jsignell — COMMENTED** [review]
This looks really good too! The only issue is consolidating on an argument name.

File: pystac/catalog.py (hunk 1 above)
Comment (thomas-maschler) [reply pystac/catalog.py:174]: Yes, correct. For example, the `pystac_client.Client` could set the fallback strategy to `APILayoutStrategy`

**Comment by thomas-maschler** [comment]
Thanks for the review, I updated the argument name.

File: pystac/catalog.py [hunk 3]
```diff
@@ -221,6 +231,19 @@ def is_relative(self) -> bool:
...
+    def _get_strategy(self, strategy: HrefLayoutStrategy | None) -> HrefLayoutStrategy:
+        if strategy is not None:
+            return strategy
+        elif self.strategy is not None:
+            return self.strategy
+        elif root := self.get_root():
+            if root.strategy is not None:
+                return root.strategy
+            else:
+                return root._fallback_strategy
```
Comment (jsignell) [inline pystac/catalog.py:243]: I think this rename is a lot clearer too :+1:

**Review by jsignell — APPROVED** [review]

**Pull request merged**